scvi/models/scanvi.py

```diff
@@ -38,202 +44,197 @@ class SCANVI(VAE):
-        px_rate = outputs["px_rate"]
-        px_dropout = outputs["px_dropout"]
-        qz1_m = outputs["qz_m"]
-        qz1_v = outputs["qz_v"]
-        z1 = outputs["z"]
-        ql_m = outputs["ql_m"]
-        ql_v = outputs["ql_v"]
-
-        # Enumerate choices of label
-        ys, z1s = broadcast_labels(y, z1, n_broadcast=self.n_labels)
-        qz2_m, qz2_v, z2 = self.encoder_z2_z1(z1s, ys)
-        pz1_m, pz1_v = self.decoder_z1_z2(z2, ys)
-
-        reconst_loss = self.get_reconstruction_loss(x, px_rate, px_r, px_dropout)
-
-        # KL Divergence
-        mean = torch.zeros_like(qz2_m)
-        scale = torch.ones_like(qz2_v)
-
-        kl_divergence_z2 = kl(
-            Normal(qz2_m, torch.sqrt(qz2_v)), Normal(mean, scale)
-        ).sum(dim=1)
-        loss_z1_unweight = -Normal(pz1_m, torch.sqrt(pz1_v)).log_prob(z1s).sum(dim=-1)
-        loss_z1_weight = Normal(qz1_m, torch.sqrt(qz1_v)).log_prob(z1).sum(dim=-1)
-        kl_divergence_l = kl(
-            Normal(ql_m, torch.sqrt(ql_v)),
-            Normal(local_l_mean, torch.sqrt(local_l_var)),
-        ).sum(dim=1)
-
-        if is_labelled:
-            return (
-                reconst_loss + loss_z1_weight + loss_z1_unweight,
-                kl_divergence_z2 + kl_divergence_l,
-                0.0,
+        if self._is_trained_base is not True:
+            self._unsupervised_trainer = UnsupervisedTrainer(
+                self._base_model,
+                self.adata,
+                train_size=train_size,
+                test_size=test_size,
+                n_iter_kl_warmup=n_iter_kl_warmup,
+                n_epochs_kl_warmup=n_epochs_kl_warmup,
+                frequency=frequency,
+                use_cuda=self.use_cuda,
+                **unsupervised_trainer_kwargs,
+            )
+            self._unsupervised_trainer.train(
+                n_epochs=n_epochs_unsupervised, lr=lr, **train_kwargs
             )
+            self._is_trained_base = True
 
-        probs = self.classifier(z1)
-        reconst_loss += loss_z1_weight + (
-            (loss_z1_unweight).view(self.n_labels, -1).t() * probs
-        ).sum(dim=1)
+        self.model.load_state_dict(self._base_model.state_dict(), strict=False)
 
-        kl_divergence = (kl_divergence_z2.view(self.n_labels, -1).t() * probs).sum(
-            dim=1
+        self.trainer = SemiSupervisedTrainer(
+            self.model, self.adata, use_cuda=self.use_cuda
         )
-        kl_divergence += kl(
-            Categorical(probs=probs),
-            Categorical(probs=self.y_prior.repeat(probs.size(0), 1)),
-        )
-        kl_divergence += kl_divergence_l
 
-        return reconst_loss, kl_divergence, 0.0
+        self.trainer.unlabelled_set = self.trainer.create_posterior(
+            indices=self._unlabeled_indices
+        )
+        self.trainer.labelled_set = self.trainer.create_posterior(
+            indices=self._labeled_indices
+        )
+        self.trainer.train(n_epochs=n_epochs_semisupervised)
```

Comment:
sometimes we adjust some other parameters in scanvi such as the n_epochs_classifier or others. It would be great to be able to access those as well. This also includes the hierarchical version of scanvi or the alpha parameter etc..

Reply:
included kwargs for semisupervised trainer and train function